contracts/compressor/IntentCompressor.sol

```diff
@@ -0,0 +1,335 @@
+     * @param token The token address
+     * @param spender The spender address
+     * @param amount The amount of tokens required
+     */
+    error InsufficientTokenAllowance(
+        address token,
+        address spender,
+        uint256 amount
+    );
+
+    address public immutable PROVER;
+    IInbox public immutable INBOX;
+    IIntentSource public immutable INTENT_SOURCE;
+
+    constructor(address _intentSource, address _inbox, address _prover) {
+        INBOX = IInbox(_inbox);
+        INTENT_SOURCE = IIntentSource(_intentSource);
+        PROVER = _prover;
+    }
+
+    function fulfill(
+        bytes32 payload,
+        bytes32 rewardHash,
+        bytes32 routeSalt
+    ) external returns (bytes[] memory) {
+        EncodedFulfillment memory encodedFulfillment = decodeFulfillPayload(
+            payload
+        );
+
+        Route memory route = _constructRoute(
+            routeSalt,
+            encodedFulfillment.sourceChainIndex,
+            encodedFulfillment.destinationChainIndex,
+            encodedFulfillment.routeTokenIndex,
+            encodedFulfillment.routeAmount
+        );
+
+        bytes32 routeHash = keccak256(abi.encode(route));
+        bytes32 intentHash = keccak256(abi.encodePacked(routeHash, rewardHash));
+
+        require(
+            route.tokens.length == 1,
+            "Cannot fulfill intent multiple tokens"
+        );
+
+        // Approve route token
+        TokenAmount memory routeToken = route.tokens[0];
+        IERC20(routeToken.token).approve(address(INBOX), routeToken.amount);
+
+        if (encodedFulfillment.proveType == 1) {
+            return
+                INBOX.fulfillHyperBatched(
+                    route,
+                    rewardHash,
+                    encodedFulfillment.claimant,
+                    intentHash,
+                    PROVER
+                );
+        } else {
+            return
+                INBOX.fulfillHyperInstant(
+                    route,
+                    rewardHash,
+                    encodedFulfillment.claimant,
+                    intentHash,
+                    PROVER
+                );
+        }
+    }
+
+    function publishTransferIntentAndFund(
+        bytes32 payload
+    ) external returns (bytes32 intentHash) {
+        EncodedIntent memory encodedIntent = decodePublishPayload(payload);
+        Intent memory intent = _constructIntent(encodedIntent);
+
+        _fundIntent(intent);
+
+        return INTENT_SOURCE.publishAndFund(intent, false);
+    }
+
+    // ======================== Public Functions ========================
+
+    function getChainIds() public pure returns (uint16[6] memory) {
+        return [1, 10, 137, 8453, 5000, 42161];
+    }
+
+    function getTokens() public pure returns (address[15] memory) {
+        return [
+            // Ethereum
+            0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48, // USDC
```

Comment:
Does it make sense to split tokens by chains? So token `0` is always USDC on any chain?

Reply:
Using a double mapping, these values will be written in storage, and accessing them will be more expensive. It probably looks cooler that way, but it may be overkill.

Reply:
That's why I said `Kinda like double mapping`. 
There is no use to have all the tokens in the array, they are from different chains and we will never loop through them. Thinking about gas efficiency, it doesn't make sense to create this whole array in memory just to be discarded.

I think the way it could be done is a function with switch cases:
```
getToken(uint8 chainId, uint8 tokenId) pure (returns address) {
    if (chainId == 1) {
      if (tokenId == 0) {
          return address(0xUSDC)
      }
  
    } else if (chainId == 10)
    ...
    }  else {
       revert Wrong();
    }
}
```

Reply:
Would make DevX much better having `USDC` as `0` on any chain

Reply:
Also allows us to have 256 token per chain instead of 256 tokens total globally.

Reply:
Added decimal to the function